runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/DAGraph.java

```diff
@@ -77,13 +85,15 @@ public void merge(DAGraph<T, U> parent) {
      * in the DAG with no dependencies.
      */
     public void prepare() {
-        for (Map.Entry<String, U> entry: graph.entrySet()) {
-            entry.getValue().prepare();
-        }
-
-        initializeQueue();
-        if (queue.isEmpty()) {
-            throw new RuntimeException("Found circular dependency");
+        if (isPreparer()) {
+            for (Map.Entry<String, U> entry : graph.entrySet()) {
```

Comment:
if looking at values only, not keys, simpler Map traversal is like this:
```java
for(ValueType value : graph.values()) {
    value.foo();
}
```
(where valueType is whatever the type of the item is)